src/Compilers/Core/Portable/SourceFileResolver.cs

```diff
@@ -87,14 +146,17 @@ public override bool Equals(object obj)
             }
 
             var other = (SourceFileResolver)obj;
-            return string.Equals(_baseDirectory, other._baseDirectory, StringComparison.Ordinal) &&
-                _searchPaths.SequenceEqual(other._searchPaths, StringComparer.Ordinal);
+            return
+                string.Equals(_baseDirectory, other._baseDirectory, StringComparison.Ordinal) &&
+                _searchPaths.SequenceEqual(other._searchPaths, StringComparer.Ordinal) &&
+                (_pathMap.IsDefaultOrEmpty ? other._pathMap.IsDefaultOrEmpty : _pathMap.SequenceEqual(other._pathMap));
         }
 
         public override int GetHashCode()
         {
             return Hash.Combine(_baseDirectory != null ? StringComparer.Ordinal.GetHashCode(_baseDirectory) : 0,
-                   Hash.CombineValues(_searchPaths, StringComparer.Ordinal));
+                   Hash.Combine(Hash.CombineValues(_searchPaths, StringComparer.Ordinal),
+                   Hash.CombineValues(_pathMap)));
```

Comment:
A very small number of source file resolvers are produced per compilation, and they are not normally placed into hash tables. We do not spend the effort to make every possible data structure as efficient as possible when that code is not on any critical performance path, and when that code would be a maintenance burden.